### PR TITLE
difficulty opcode implementation

### PR DESCRIPTION
**Pull request by akinovak**
- circuit&additional types

### CODE REVIEW

File: zkevm-circuits/src/evm_circuit/param.rs [hunk 1]
```diff
@@ -36,3 +36,6 @@ pub(crate) const N_BYTES_GAS: usize = N_BYTES_U64;
 
 // Number of bytes that will be used for call data's size.
 pub(crate) const N_BYTES_CALLDATASIZE: usize = N_BYTES_U64;
+
+// Number of bytes that will be used for block difficulty.
+pub(crate) const N_BYTES_DIFFICULTY: usize = N_BYTES_U64;
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/param.rs:41]: this is not used?

**Review by icemelon — COMMENTED** [review]
Thanks for the contribution! Could you also add the spec for `DIFFICULTY` opcode at https://github.com/appliedzkp/zkevm-specs?

File: bus-mapping/src/evm/opcodes.rs [hunk 2]
```diff
@@ -127,6 +127,7 @@ fn fn_gen_associated_ops(opcode_id: &OpcodeId) -> FnGenAssociatedOps {
         OpcodeId::COINBASE => StackOnlyOpcode::<0, 1>::gen_associated_ops,
         OpcodeId::TIMESTAMP => StackOnlyOpcode::<0, 1>::gen_associated_ops,
         OpcodeId::NUMBER => StackOnlyOpcode::<0, 1>::gen_associated_ops,
+        OpcodeId::DIFFICULTY => StackOnlyOpcode::<0, 1>::gen_associated_ops,
         // OpcodeId::DIFFICULTY => {},
```
[the remark this answers is not in the document]
Comment (akinovak) [reply bus-mapping/src/evm/opcodes.rs:131]: my test is failing after this line removal

**Comment by akinovak** [comment]
> Thanks for the contribution! Could you also add the spec for `DIFFICULTY` opcode at https://github.com/appliedzkp/zkevm-specs?

I see https://github.com/appliedzkp/zkevm-specs/blob/master/specs/opcode/41COINBASE-45GASLIMIT_48BASEFEE.md block ctx spec here.

File: bus-mapping/src/evm/opcodes.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (icemelon) [reply bus-mapping/src/evm/opcodes.rs:131]: I mean remove the commented line.

File: zkevm-circuits/src/evm_circuit/execution/difficulty.rs [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        self.same_context.assign_exec_step(region, offset, step)?;
+
+        let difficulty = block.rws[step.rw_indices[0]].stack_value();
+
+        self.difficulty.assign(
+            region,
+            offset,
+            Some(u64::try_from(difficulty).unwrap().to_le_bytes()),
```
Comment (roynalnaruto) [inline zkevm-circuits/src/evm_circuit/execution/difficulty.rs:76]: You may add this if you feel like, but in my opinion this looks cleaner:

```suggestion
            Some(difficulty.as_u64().to_le_bytes()),
```

**Review by roynalnaruto — COMMENTED** [review]
Just a tiny nitpick.

Also, you will need to create a PR to the Specs repo: https://github.com/appliedzkp/zkevm-specs.

Once the spec is merged, only then the circuit PR will be reviewed :)

**Comment by kobigurk** [comment]
> Just a tiny nitpick.
> 
> Also, you will need to create a PR to the Specs repo: https://github.com/appliedzkp/zkevm-specs.
> 
> Once the spec is merged, only then the circuit PR will be reviewed :)

Doesn't https://github.com/appliedzkp/zkevm-specs/blob/537c791e073e4357e42434e82910443328c80902/src/zkevm_specs/evm/execution/block_ctx.py#L25 cover this already? Besides the concern about the size mentioned here https://github.com/appliedzkp/zkevm-specs/issues/186

**Comment by roynalnaruto** [comment]
> Doesn't https://github.com/appliedzkp/zkevm-specs/blob/537c791e073e4357e42434e82910443328c80902/src/zkevm_specs/evm/execution/block_ctx.py#L25 cover this already?

@kobigurk My bad! This does cover all block context fields 👍

**Comment by kobigurk** [comment]
> > Doesn't https://github.com/appliedzkp/zkevm-specs/blob/537c791e073e4357e42434e82910443328c80902/src/zkevm_specs/evm/execution/block_ctx.py#L25 cover this already?
> 
> @kobigurk My bad! This does cover all block context fields 👍

No worries at all, we're still learning :) Curious what you think about the size!

**Comment by roynalnaruto** [comment]
> Curious what you think about the size!

I would be in favour of aligning our codebase with `go-ethereum`, and limiting `DIFFICULTY` to 10 bytes.

**Comment by kobigurk** [comment]
> > Curious what you think about the size!
> 
> I would be in favour of aligning our codebase with `go-ethereum`, and limiting `DIFFICULTY` to 10 bytes.

BTW, about the spec - I see that the first three fields are not mentioned to be RLC, although the implementation uses `rlc_to_fq`: https://github.com/appliedzkp/zkevm-specs/blob/537c791e073e4357e42434e82910443328c80902/src/zkevm_specs/evm/execution/block_ctx.py#L14

**Comment by roynalnaruto** [comment]
BTW, stepping back a bit, the `BlockCtxGadget` has already been implemented:
https://github.com/appliedzkp/zkevm-circuits/blob/main/zkevm-circuits/src/evm_circuit/execution/block_ctx.rs

And it includes tests for `DIFFICULTY` opcode as well (although `difficulty` is 32-bytes in the current implementation).

In that case #459 should not be required, unless I'm missing something.

**Comment by roynalnaruto** [comment]
> about the spec - I see that the first three fields are not mentioned to be RLC

WRT, I think the specs need to be updated. Since those opcodes result in a stack push, the values will be in the RLC form. And since the spec specifies only some of them as `RLC`, it is a bit misleading that the other fields are not `RLC`.

**Comment by kobigurk** [comment]
> BTW, stepping back a bit, the `BlockCtxGadget` has already been implemented: https://github.com/appliedzkp/zkevm-circuits/blob/main/zkevm-circuits/src/evm_circuit/execution/block_ctx.rs
> 
> And it includes tests for `DIFFICULTY` opcode as well (although `difficulty` is 32-bytes in the current implementation).
> 
> In that case #459 should not be required, unless I'm missing something.

😂  yes, it seems like it!

Maybe we'll do a brush over this and update the spec like you suggest.

**Pull request closed without merging**